smartsuspend.lua

```diff
@@ -0,0 +1,117 @@
+    return building ~= nil
+        and df.building_constructionst:is_instance(building)
+        and (building.type == df.construction_type.Wall or building.type == df.construction_type.Fortification)
+end
+
+--- True if there is a construction plan to build an unwalkable tile
+---@param pos coord
+---@return boolean
+local function hasUnwalkableConstruction(pos)
+    --- @type building_constructionst|building
+    local building = dfhack.buildings.findAtTile(pos)
+    return isBlockingConstruction(building)
+end
+
+--- Check if the tile can be walked on
+---@param pos coord
+local function walkable(pos)
+    local tt = dfhack.maps.getTileType(pos)
+    if not tt then
+        return false
+    end
+    local attrs = df.tiletype.attrs[tt]
+    local shape_attrs = df.tiletype_shape.attrs[attrs.shape]
+    return shape_attrs.walkable
+end
+
+--- List neighbour coordinates of a position
+---@param pos coord
+---@return table<number, coord>
+local function neighbours(pos)
+    return {
+        {x=pos.x-1, y=pos.y, z=pos.z},
+        {x=pos.x+1, y=pos.y, z=pos.z},
+        {x=pos.x, y=pos.y-1, z=pos.z},
+        {x=pos.x, y=pos.y+1, z=pos.z},
+    }
+end
+
+
+--- Get the amount of risk a tile is to be blocked
+--- -1: There is a nearby walkable area with no plan to build a wall
+--- >=0: Surrounded by either unwalkable tiles, or tiles that will be constructed
+--- with unwalkable buildings. The value is the number of already unwalkable tiles.
+---@param pos coord
+local function riskOfStuckConstructionAt(pos)
+    local risk = 0
+    for _,neighbourPos in pairs(neighbours(pos)) do
+        if not walkable(neighbourPos) then
+            -- blocked neighbour, increase danger
+            risk = risk + 1
+        elseif not hasUnwalkableConstruction(neighbourPos) then
+            -- walkable neighbour with no plan to build a wall, no danger
+            return -1
+        end
+    end
+    return risk
+end
+
+--- Return true if this job is at risk of blocking another one
+function jobShouldBeSuspended(job)
+    local building = getBuildingOfJob(job)
+    --- Not building a wall, no risk
+    if not building or not isBlockingConstruction(building) then return false end
+
+    --- job.pos is sometimes off by one, get the building pos
+    local pos = {x=building.centerx,y=building.centery,z=building.z}
+
+    --- Get self risk of being blocked
+    local risk = riskOfStuckConstructionAt(pos)
+
+    for _,neighbourPos in pairs(neighbours(pos)) do
+        if hasUnwalkableConstruction(neighbourPos) and riskOfStuckConstructionAt(neighbourPos) > risk then
+            --- This neighbour job is at greater risk of getting stuck
+            return true
+        end
+    end
+
+    return false
+end
+
+if dfhack_flags.module then
+    return
+end
+
+for _,job in utils.listpairs(df.global.world.jobs.list) do
+    if jobShouldBeSuspended(job) then
+        suspend(job)
+    else
+        job.flags.suspend = false
```

Comment:
this script shouldn't be unsuspending jobs unless it's taking over the job of autounsuspend (which I recommend that it does -- I think this functionality should be part of `autounsuspend`, on by default, and controllable with a setting)

Reply:
Yep, it makes sense